### DISCUSSION TITLE
How to save the failed_at field as ISODate?

### DISCUSSION THREAD

**Discussion by BehroozBvk** [jenssegers/laravel-mongodb#2605] [Show and tell]
Hi, Before telling the solution, I must say that if the configuration considered for `failed.driver` in the `queue.php` file is `database-uuids` by default, the value of the `failed_at` field in Mongo is an empty object and no date is stored in it.

But if we change the driver to **mongodb**, you will see that the date is stored as a **string** of the `timestamp` value in the `failed_at` field. And this was not what I expected and I wanted the date to be stored as an **ISODate**.

However, to save the date in the `failed_at` field as **ISODate**, you can take the following steps.

Step 1:
-------
Run the following command 
```bash
php artisan make:provider CustomMongodbQueueServiceProvider
```
Then rewrite the `CustomMongodbQueueServiceProvider.php` file as below:

```php
<?php

namespace App\Providers;

use Jenssegers\Mongodb\MongodbQueueServiceProvider;

class CustomMongodbQueueServiceProvider extends MongodbQueueServiceProvider
{
    /**
     * Create a new MongoDB failed job provider.
     */
    protected function mongoFailedJobProvider(array $config): CustomMongoFailedJobProvider
    {
        return new CustomMongoFailedJobProvider($this->app['db'], $config['database'], $config['table']);
    }
}
```

Step 2:
-------
Run the following command :
```bash
php artisan make:provider CustomMongoFailedJobProvider
```
Then rewrite the `CustomMongoFailedJobProvider.php` file as below:

```php
<?php

namespace App\Providers;

use Jenssegers\Mongodb\Queue\Failed\MongoFailedJobProvider;

class CustomMongoFailedJobProvider extends MongoFailedJobProvider
{
    public function log($connection, $queue, $payload, $exception): void
    {
        $failed_at = new \MongoDB\BSON\UTCDateTime(now());
        $exception = (string) $exception;
        $this->getTable()->insert(compact('connection', 'queue', 'payload', 'failed_at', 'exception'));
    }
}
```

I think it is generally clear what we did, a simple override of two service providers related to `MongoFailedJobProvider`, `MongodbQueueServiceProvider` through which we can rewrite the `log` method according to our needs.

Step 3:
-------
Now we have to register our service provider in the `config/app.php` file :
```php
'providers' => [
    // Other Service Providers ...
    App\Providers\CustomMongodbQueueServiceProvider::class,
],
```

Step 4:
-------
Then run the following command :
```bash
composer dump-autoload
```

Well, our work is finished, now you can test the result. I hope this solution helped you.

![photo_2023-09-06_17-01-47](https://github.com/mongodb/laravel-mongodb/assets/17563730/3bd3961a-f45b-4c6c-ae69-462533463296)

**Comment by GromNaN** [comment]
Reading the laravel code, I see that the type of the `failed_at` field depends on the database type. With SQL databases, it's a [datetime](https://github.com/laravel/framework/blob/52b301797d7a229bf7ebd3e2eefc762487738578/src/Illuminate/Queue/Failed/DatabaseFailedJobProvider.php#L56-L65). With DynamoDB, it's a [timestamp](https://github.com/laravel/framework/blob/52b301797d7a229bf7ebd3e2eefc762487738578/src/Illuminate/Queue/Failed/DynamoDbFailedJobProvider.php#L73). For MongoDB, it was implemented in #792 with a timestamp.

https://github.com/mongodb/laravel-mongodb/blob/9bc8999b79279e46bc92a4e1951a0dc70ce2d63a/src/Queue/Failed/MongoFailedJobProvider.php#L25-L34

I agree that an UTCDateTime should be used. Would you like to submit a PR?

Also, it looks like the [Job Batching](https://github.com/laravel/framework/pull/32830) feature implemented in Laravel 8.x needs some update.

https://jira.mongodb.org/browse/PHPORM-87

**Reply by BehroozBvk** [reply]
yes I agree.
